### ISSUE TITLE
Crashes when running Regent on updated CI

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1297]
I'm hoping to get some help debugging an issue that occurs in the updated CI. This is technically a part of #1295 but I'm splitting it off to keep the discussion clean.

My understanding of the bug so far is:

  * Only Regent tests fail. (I think I've fixed all the Legion test failures.)
  * MPI/GASNet is not required.
  * I can reproduce in GCC or Clang.
  * I can reproduce in any C++ standard version.
  * I can reproduce in debug or release mode, though I often don't get useful backtraces in either case.
  * I can reproduce in Ubuntu 20.04 or 18.04. I can reproduce with Terra built against LLVM 12 or 10. Really, any newly built image I've tried has failed (though so far I have yet to go back to Ubuntu 16.04).
  * Failures occur non-deterministically, but often enough that the CI fails most of the time. E.g., in my last full CI run, 16 jobs failed out of 138. Reproducing by hand is harder, but possible with `gitlab2docker.py`.
  * However, I have yet to reproduce with a direct invokation of a failing command, even when that command just failed in the test suite.
  * So far, I have only reproduced on the Sapling `n` nodes. A disproportionate number of failures happen on `n0002`, though some failures occur on `n0000` and `n0003`. However, it's hard to be definitive because the failures are non-deterministic.

Here's an example failure in CI: https://gitlab.com/StanfordLegion/legion/-/jobs/2712540417

To see the rest of the failures, go to: https://gitlab.com/StanfordLegion/legion/-/pipelines/586293418/failures

Here's a sample `gitlab2docker.py` command to reproduce:

```
python3 tools/gitlab2docker.py -b gitlab-ci-docker-update .gitlab-ci.yml gcc9_cxx17_debug_regent -n
```

Then inside the resulting Docker container, run `/script.sh`.

Using this approach, I got the following failure:

```
[FAIL] (run_pass) (signal SIGSEGV) tests/regent/run_pass/parallelize_tasks2.rg
parallelize_tasks2.rg -ll:cpu 4 -fbounds-checks 1 -fparallelize-dop 5 -fjobs 1 -level runtime=5,threads=5
[0x7f7810b11d54]
[0x10660]
[0xc01]
[0xc01]
/repo/bindings/regent/libregent.so(+0x3015c01) [0x7f77dd000c01]
/repo/bindings/regent/libregent.so(+0x3015c01) [0x7f77dd000c01]
```

However, I am NOT able to reproduce when I run this command by hand:

```
i=0; while REALM_FREEZE_ON_ERROR=1 ./regent.py tests/regent/run_pass/parallelize_tasks2.rg -ll:cpu 4 -fbound
s-checks 1 -fparallelize-dop 5; do let i++; echo $i; done
```

I'm at 100 iterations now and no crash in sight.

Any ideas? I'm sort of tearing my hair out at this point.

**Comment by lightsighter** [comment]
Usually when I have a hard to reproduce bug, I'll use gitlab2docker.py on one of the n000* nodes (which you've already done). Then I'll actually compile the regent program into a binary using `SAVEOBJ=1`. Then I'll run it repeatedly, often in parallel with a full run of the test suite to generate a bunch of noise on the same node. 

The lack of a backtrace suggests to me that the segfault is either coming at start-up or shutdown, either before Realm has registered its signal handlers or after it has unregistered them.

**Comment by lightsighter** [comment]
Another thing to try is to set `REALM_FREEZE_ON_ERROR=1` and then watch the test suite run and look for processes that are not exiting in a reasonable amount of time and then you can attach and look at them up until the timeout kills them.

**Comment by eddy16112** [comment]
What I did before was to modify the `travis.py` to disable the compilation, `test.py` to remove the tests that have no issues, and run `REALM_FREEZE_ON_ERROR=1 ./script.sh` until I caught the crash.

**Comment by lightsighter** [comment]
I've got a debugger on some processes that segfaulted but they are not particularly helpful:

```
#0  0x00007f0b2bebd3bf in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, 
    req=req@entry=0x7f0b3373caa0, rem=rem@entry=0x7f0b3373caa0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007f0b2bec3047 in __GI___nanosleep (requested_time=requested_time@entry=0x7f0b3373caa0, 
    remaining=remaining@entry=0x7f0b3373caa0) at nanosleep.c:27
#2  0x00007f0b2bec2f7e in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x00007f0b2a67813b in Realm::realm_freeze (signal=6) at /repo/runtime/realm/runtime_impl.cc:179
#4  <signal handler called>
#5  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#6  0x00007f0b2be02859 in __GI_abort () at abort.c:79
#7  0x00007f0b2c416d65 in ?? ()
#8  0x00007f0af0200c80 in ?? ()
#9  0x00000000f0200d18 in ?? ()
#10 0x00007f0af0202e10 in ?? ()
#11 0x00007f0b337e53a0 in ?? ()
#12 0x0000000000000000 in ?? ()
```

**Comment by lightsighter** [comment]
Interestingly, this is not a Realm thread. I have no idea where it came from.

**Comment by elliottslaughter** [comment]
I just got a crash after 167 iterations. `REALM_FREEZE_ON_ERROR=1` did not catch it.

I'm guessing this is somewhere in LuaJIT? Still not especially helpful.

**Comment by lightsighter** [comment]
So interestingly, all my failures are calling `abort` and not segfaulting. It's definitely a non-Realm thread though. Since these are all self-contained Regent programs, I'm guessing it's LuaJIT making this thread. How do I get a debug build of LuaJIT?

**Comment by elliottslaughter** [comment]
I rebuilt the image last night to update LuaJIT to something more recent. This did not fix the problem, though the symptoms seem slightly different now.

Next I set up a debug build of Terra/LuaJIT. I've confirmed that Terra has debug symbols, but the backtrace is still useless.

```
$ i=0; while LAUNCHER="gdb -ex run -ex quit --args" ./regent.py tests/regent/run_pass/region_create_ispace.rg; do let i++; echo $i; done

...

Program received signal SIGSEGV, Segmentation fault.
0x000055553080fbd8 in ?? ()
A debugging session is active.

        Inferior 1 [process 6198] will be killed.

Quit anyway? (y or n) n
Not confirmed.
(gdb) bt
#0  0x000055553080fbd8 in ?? ()
#1  0x00007ffff7941970 in ?? ()
#2  0x00007fffffffe210 in ?? ()
#3  0x0000555556773ce0 in ?? ()
#4  0x0000001000000000 in ?? ()
#5  0x00007ffff7951380 in ?? ()
#6  0x00007ffff7941954 in ?? ()
#7  0x0000000000000000 in ?? ()
```

I see there are some debugging options for LuaJIT, so I guess I'll try those next:

https://github.com/LuaJIT/LuaJIT/blob/6053b04815ecbc8eec1e361ceb64e68fb8fac1b3/src/Makefile#L118-L152

**Comment by elliottslaughter** [comment]
I've confirmed the error occurs only in Terra built with Make. The errors go away in CMake in every configuration I've tested. I'm not sure what the difference is yet, but this potentially gives us a way to move forward.

**Comment by elliottslaughter** [comment]
So much for my confirmation. The [new jobs still fail](https://gitlab.com/StanfordLegion/legion/-/jobs/2719159623).

I had 5 iterations of `/script.sh` pass in my old container, and now in the new container it fails half a dozen tests on my first try. Maybe it's sensitive to the amount of load?

**Comment by elliottslaughter** [comment]
I have run nearly 180 iterations of `/script.sh` on a `g` node and have yet to see a failure. This is with 3x oversubscription (i.e., I had three concurrent copies of `/script.sh` going, and each was set to use all the cores on the machine, meaning were 3x the number of active threads as cores/hyperthreads on the machine).

My best guess at this point is we're seeing a sensitivity to the host OS version. Docker does not hide the kernel from the userspace process. And our CI hosts are on Ubuntu 14.04, so it's quite possible we're simply not up-to-date enough to run Ubuntu 20.04 containers seamlessly.

Since this bug has been so hard to replicate, I am planning to run the CI on our newer `c` nodes for a short time. If that works, I'll look into getting the `n` nodes upgraded, which will hopefully put this one behind us.

**Comment by elliottslaughter** [comment]
I moved the CI to our `c` nodes temporarily. My observations:

 * I no longer see any single-node segfaults at all. This failure mode seems to have been resolved.
 * I do see multi-node failures, but the failure mode seems different. The failure rate is nearly 100%, suggesting something that should (hopefully) be easy to debug. I have yet to look into it though.
 * I see some new failures in the profiler tests, differences between Python and Rust output. This happens on `master` so it's not specific to my CI changes. Haven't had a chance to look at this yet, but I'm guessing it's a timing bug that has shaken loose.
 * `master` is otherwise passing and I've done a few full runs to be sure everything is fine.

Overall it seems like a step in the right direction, so I've asked the CS staff to upgrade the `n` nodes to Ubuntu 20.04.

**Comment by lightsighter** [comment]
> I see some new failures in the profiler tests, differences between Python and Rust output. This happens on master so it's not specific to my CI changes. Haven't had a chance to look at this yet, but I'm guessing it's a timing bug that has shaken loose.

Also happening on the control replication branch as well.

**Comment by elliottslaughter** [comment]
For posterity, here is a set of "bad" and "good" files from the `master` branch:

 * "bad" (Rust and Python output differ): [prof_0.gz](https://github.com/StanfordLegion/legion/files/9124695/prof_0.gz)
 * "good" (Rust and Python output match): [good_prof_0.gz](https://github.com/StanfordLegion/legion/files/9124698/good_prof_0.gz)
 
The test in these examples is https://gitlab.com/StanfordLegion/legion/-/blob/master/language/tests/regent/run_pass/annotations.rg. It is definitely a timing related issue as I needed to run an interfering job to make it repro.

The profilers themselves are deterministic so it's hopefully just a matter of walking through the logic to figure out what's wrong.

**Comment by elliottslaughter** [comment]
This has hopefully been fixed by https://gitlab.com/StanfordLegion/legion/-/commit/457b5835217ff5837fbfe2fb4a20acde6e01cd20

**Comment by elliottslaughter** [comment]
I'm looking at the last Regent crash now. Example failure: https://gitlab.com/StanfordLegion/legion/-/jobs/2724972395

Example of a failing command line (after running `/script.sh`; also I built a debug Terra): `LAUNCHER="mpirun -n 1 gdb --args" ./regent.py ../tutorial/00_hello_world/hello_world.rg`

There is one thread in the final stages of shutdown. Backtrace:

```
Thread 1 "terra" received signal SIGSEGV, Segmentation fault.
0x00007f934c96b9f0 in ?? ()
(gdb) bt
#0  0x00007f934c96b9f0 in ?? ()
#1  0x00007f934d7328e9 in __run_exit_handlers (status=0, 
    listp=0x7f934d8d8718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true)
    at exit.c:91
#2  0x00007f934d732a60 in __GI_exit (status=<optimized out>) at exit.c:139
#3  0x00007f934d71008a in __libc_start_main (main=
    0xbadef0 <main(int, char**)>, argc=2, argv=0x7ffcf44a6328, 
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7ffcf44a6318) at ../csu/libc-start.c:342
#4  0x0000000000badcee in _start ()
```

This appears to be a crash in LuaJIT (hence the lack of useful debug info, even with a debug build). But my attempts to enable LuaJIT assertions (`-DLUA_USE_APICHECK -DLUA_USE_ASSERT`, the highest level that LuaJIT appears to support) doesn't complain about anything.

I have noticed that this crash appears to be Clang-specific. E.g., an equivalent GCC job passes: https://gitlab.com/StanfordLegion/legion/-/jobs/2724972314 . Not sure that leads to a next debugging step though, aside from turning the test off.

My only other thought is to go back to OpenMPI and see if that might play nicer with all the pieces.

**Comment by elliottslaughter** [comment]
Some updates:

I tried Clang 14 with Ubuntu 22.04 and it doesn't make any difference. I have a test going for Ubuntu 18.04 in case it's sensitive to that.

I went looking for possible calls to `atexit`, since we seem to be in an exit handler in the crash. Running in gdb with `b atexit` breaks in one location, here: https://gitlab.com/StanfordLegion/legion/-/blob/457b5835217ff5837fbfe2fb4a20acde6e01cd20/runtime/realm/logging.cc#L383

However, commenting this code out does *not* fix the issue (and I confirmed the `atexit` call is gone in gdb).

My only other thought is that maybe another library is setting an exit handler: LLVM, MPI, etc.

**Comment by lightsighter** [comment]
Here is the backtrace of a crash:

```
#0  0x00007f1f0c0599f0 in ?? ()
#1  0x00007f1f0ce208e9 in __run_exit_handlers (status=0, 
    listp=0x7f1f0cfc6718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true)
    at exit.c:91
#2  0x00007f1f0ce20a60 in __GI_exit (status=<optimized out>) at exit.c:139
#3  0x00007f1f0cdfe08a in __libc_start_main (main=0x556b984bdfc3 <main>, 
    argc=2, argv=0x7ffe87c02928, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7ffe87c02918)
    at ../csu/libc-start.c:342
#4  0x0000556b984bdd0e in _start ()
```

So the bad function pointer is `7f1f0c0599f0`, i dumped a backtrace for every single call into the `__new_exitfn` in libc and here is the backtrace that matches with that function pointer:

```
Breakpoint 2, __new_exitfn (listp=0x7f1f0cfc6718 <__exit_funcs>)
    at cxa_atexit.c:82
82	in cxa_atexit.c
#0  __new_exitfn (listp=0x7f1f0cfc6718 <__exit_funcs>) at cxa_atexit.c:82
#1  0x00007f1f0ce20aa0 in __on_exit (func=0x7f1f0c0599f0 <gasneti_on_exit>, 
    arg=0x0) at on_exit.c:34
#2  0x00007f1f0c04f2c1 in gasnetc_attach_primary ()
   from /repo/bindings/regent/libregent.so
#3  0x00007f1f0c04f783 in gex_Client_Init_GASNET_202230PARnopshmFASTnodebugnotracenostatsnodebugmallocnosrclines () from /repo/bindings/regent/libregent.so
#4  0x00007f1f0ba75b8b in Realm::GASNetEXInternal::init (this=0x556ba1600b00, 
    argc=0x7ffe87bf99e4, argv=0x7ffe87bf99d8)
    at /repo/runtime/realm/gasnetex/gasnetex_internal.cc:3176
#5  0x00007f1f0ba6706c in Realm::GASNetEXModule::create_network_module (
    runtime=0x556ba16b5f60, argc=0x7ffe87bf99e4, argv=0x7ffe87bf99d8)
    at /repo/runtime/realm/gasnetex/gasnetex_module.cc:470
#6  0x00007f1f0b763f69 in Realm::ModuleRegistrar::NetworkRegistration<Realm::GASNetEXModule>::create_network_module (this=0x7f1f0ca634f8 <registration_88>, 
    runtime=0x556ba16b5f60, argc=0x7ffe87bf99e4, argv=0x7ffe87bf99d8)
    at /repo/runtime/realm/module.h:174
#7  0x00007f1f0ba95e52 in Realm::ModuleRegistrar::create_network_modules (
    this=0x556ba16b7118, modules=std::vector of length 0, capacity 0, 
    argc=0x7ffe87bf99e4, argv=0x7ffe87bf99d8)
    at /repo/runtime/realm/network.cc:593
#8  0x00007f1f0b60a505 in Realm::RuntimeImpl::network_init (
    this=0x556ba16b5f60, argc=0x7ffe87c01c64, argv=0x7ffe87c01c68)
    at /repo/runtime/realm/runtime_impl.cc:1022
#9  0x00007f1f0b60a42a in Realm::Runtime::network_init (this=0x7ffe87c01b88, 
    argc=0x7ffe87c01c64, argv=0x7ffe87c01c68)
    at /repo/runtime/realm/runtime_impl.cc:429
#10 0x00007f1f0b0969d4 in Legion::Internal::Runtime::initialize (
    argc=0x7ffe87c01c64, argv=0x7ffe87c01c68, filter=true)
    at /repo/runtime/legion/runtime.cc:22359
#11 0x00007f1f0aad354c in Legion::Runtime::initialize (argc=0x7ffe87c01c64, 
    argv=0x7ffe87c01c68, filter=true) at /repo/runtime/legion/legion.cc:6938
#12 0x00007f1f0ab2158c in legion_runtime_initialize (argc=0x7ffe87c01c64, 
    argv=0x7ffe87c01c68, filter=true) at /repo/runtime/legion/legion_c.cc:7171
#13 0x00007f1f099c8ff3 in $main ()
#14 0x00007f1f099c5032 in $wrapper ()
#15 0x0000556b9c17865d in lj_vm_ffi_call ()
#16 0x0000556b9c135d41 in lj_ccall_func ()
#17 0x0000556b9c11ba50 in lj_cf_ffi_meta___call ()
#18 0x0000556b9c176246 in lj_BC_FUNCC ()
#19 0x0000556b9c12737f in lua_pcall ()
#20 0x0000556b984bebb6 in docall(lua_State*, int, int) ()
#21 0x0000556b984be1bf in main ()
```

So the issue is definitely somewhere in GASNet. I did a run with a breakpoint set on `dlclose` and it didn't look like `librealm.so` (which statically links GASNet) had been unloaded next. Now trying to use a debug build of GASNet to see what is going wrong in this function.

**Comment by lightsighter** [comment]
It's not GASNet's fault. `libregent.so` is definitely being unmapped before GASNet's `gasneti_on_exit` function is being called (note that GASNet is statically linked into `libregent.so`). I can capture the backtrace for the registration of `gasneti_on_exit` now, it looks like this:

```
(gdb) p gasneti_on_exit
$1 = {void (int, void *)} 0x7f71e19c0a10 <gasneti_on_exit>
```

When we do that, we can see that `libregent.so` is mapped here:

```
(gdb) info proc mappings
process 45266
Mapped address spaces:

          Start Addr           End Addr       Size     Offset objfile
          ...
      0x7f71df732000     0x7f71e038f000   0xc5d000        0x0 /repo/bindings/regent/libregent.so
      0x7f71e038f000     0x7f71e1b14000  0x1785000   0xc5d000 /repo/bindings/regent/libregent.so
      0x7f71e1b14000     0x7f71e245b000   0x947000  0x23e2000 /repo/bindings/regent/libregent.so
      0x7f71e245b000     0x7f71e245c000     0x1000  0x2d29000 /repo/bindings/regent/libregent.so
      0x7f71e245c000     0x7f71e24a9000    0x4d000  0x2d29000 /repo/bindings/regent/libregent.so
      0x7f71e24a9000     0x7f71e24af000     0x6000  0x2d76000 /repo/bindings/regent/libregent.so
```

You can confirm for yourself that the pointer is in the second range, so all good on that front. We then continue the execution and segfault when running the exit routines.

```
#0  0x00007f71e19c0a10 in ?? ()
#1  0x00007f71e286d8e9 in __run_exit_handlers (status=0, listp=0x7f71e2a13718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:91
#2  0x00007f71e286da60 in __GI_exit (status=<optimized out>) at exit.c:139
#3  0x00007f71e284b08a in __libc_start_main (main=0x55b439aaefc3 <main>, argc=2, argv=0x7ffeb0126778, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffeb0126768) at ../csu/libc-start.c:342
#4  0x000055b439aaed0e in _start ()
```

The pointer is exactly what we expect it to be, so we run `info proc mappings` again, and all the `libregent.so` mappings are gone! Going to try to find out where they went now.

**Comment by lightsighter** [comment]
Alright, and here is our smoking gun, LLVM being unhelpful and unloading `libregent.so` before `gasneti_on_exit` can be called:

```
warning: Temporarily disabling breakpoints for unloaded shared library "/repo/bindings/regent/libregent.so"
Stopped due to shared library event:
  Inferior unloaded /repo/bindings/regent/libregent.so
    /lib/x86_64-linux-gnu/libatomic.so.1
    /lib/x86_64-linux-gnu/libmpich.so.12
    /lib/x86_64-linux-gnu/libmpichcxx.so.12
(gdb) bt
#0  _dl_close_worker (map=map@entry=0x5605b99bf1d0, force=force@entry=false) at dl-close.c:811
#1  0x00007f7900832f05 in _dl_close_worker (force=false, map=0x5605b99bf1d0) at dl-close.c:146
#2  _dl_close (_map=0x5605b99bf1d0) at dl-close.c:859
#3  0x00007f790033f928 in __GI__dl_catch_exception (exception=exception@entry=0x7ffdb913ece0, 
    operate=operate@entry=0x7f790080b420 <dlclose_doit>, args=args@entry=0x5605b99bf1d0) at dl-error-skeleton.c:208
#4  0x00007f790033f9f3 in __GI__dl_catch_error (objname=objname@entry=0x5605b9970900, errstring=errstring@entry=0x5605b9970908, 
    mallocedp=mallocedp@entry=0x5605b99708f8, operate=operate@entry=0x7f790080b420 <dlclose_doit>, args=args@entry=0x5605b99bf1d0)
    at dl-error-skeleton.c:227
#5  0x00007f790080bb59 in _dlerror_run (operate=operate@entry=0x7f790080b420 <dlclose_doit>, args=0x5605b99bf1d0) at dlerror.c:170
#6  0x00007f790080b468 in __dlclose (handle=<optimized out>) at dlclose.c:46
#7  0x00005605b17634bd in llvm::object_deleter<llvm::sys::DynamicLibrary::HandleSet>::call(void*) ()
#8  0x00005605b17060f9 in llvm::llvm_shutdown() ()
#9  0x00005605b1636d24 in terra_llvmshutdown ()
#10 0x00005605b1509233 in main ()
```

Fuck LLVM!

**Comment by elliottslaughter** [comment]
What I'm getting from this is that you can't `terralib.linklibrary` with any library that uses `atexit`. That seems unfortunate but I don't really see a way around it, unless we call `terra_llvmshutdown()` itself as an `atexit` handler, which I'm not sure will actually do what we want (and will subject us to, e.g., `atexit` handler ordering issues, which as I understand is not well-defined).

What happened when you set `GASNET_CATCH_EXIT=0`? It seems like that may be our best approach.

I guess the other option is to make a `terra_regent` that links Legion directly so that we don't need to involve LLVM in the process of loading Legion, making the LLVM shutdown order irrelevant.

**Comment by elliottslaughter** [comment]
To answer my own question, I get 4 failures (out of 1200 tests), and an example looks like:

```
[FAIL] (run_pass) (return code 1) tests/regent/run_pass/phase_barrier.rg
phase_barrier.rg -ll:cpu 2 -fjobs 1 -level runtime=5,threads=5

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 223881 RUNNING AT abf9407aec9f
=   EXIT CODE: 1
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Hangup (signal 1)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

So I suppose the GASNet exit handler may actually be aiding the shutdown process in some way, though it's odd that we can't exit cleanly without it (these should be passing tests, not failures).

**Comment by lightsighter** [comment]
> So I suppose the GASNet exit handler may actually be aiding the shutdown process in some way, though it's odd that we can't exit cleanly without it (these should be passing tests, not failures).

Right, this is `mpirun` complaining because GASNet doesn't shutdown fast enough for its liking. The GASNet exit handler helps to accelerate the completion of all the processes in a way that is more consistent with `mpirun`'s expectations. The signal 1 error is not a failure, it just says that the something "hungup" and took longer than expected.

What exactly is `llvm_shutdown` doing? Why isn't it just shutting down `llvm` and nothing else? I don't understand why `llvm` thinks its ok to unload shared objects that it didn't load, depend on, or make during execution.

**Comment by lightsighter** [comment]
> I suppose we could call dlopen ourselves. I'm not sure what the consequences of that would be.

We should do that. `dlopen` and `dlclose` work like reference counting on the shared objects and they automatically deduplicate so things are only loaded once. We should `dlopen` `libregent.so` ourselves to ensure it survives all the way through to the end of execution.

**Comment by elliottslaughter** [comment]
Using `dlopen` seems to fix this.

The only outstanding issue now is that the HTR external test is failing: https://gitlab.com/StanfordLegion/legion/-/jobs/2742944455

**Comment by lightsighter** [comment]
Do we have any idea why that is failing? It's hard for me to explain a configuration error. @mariodirenzo can you provide any insight into what is going wrong in this failing test case?

https://gitlab.com/StanfordLegion/legion/-/jobs/2742944455

**Comment by mariodirenzo** [comment]
It apparently looks like a corruption of a data structure that is passed by value across a couple of tasks. I have already written in the slack channel with Elliott a possible debug strategy.

**Comment by elliottslaughter** [comment]
So far I have been unable to reproduce locally. Instead I hit this compile error:

```
/repo/runtime/realm/dynamic_templates.inl:215:55: error: ‘demux<Realm::DynamicTemplates::ListProduct2<L1, L2>::DemuxHelper2<TARGET, T1>, A1>’ is not a member of ‘Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >’
/repo/runtime/realm/dynamic_templates.inl: In instantiation of ‘static void Realm::DynamicTemplates::ListProduct2<L1, L2>::DemuxHelper1<TARGET>::demux(Realm::DynamicTemplates::TagType, A1) [with T1 = Realm::DynamicTemplates::Int<3>; A1 = int*; TARGET = Legion::Internal::NT_TemplateHelper::DimHelper; L1 = Realm::DynamicTemplates::IntList<1, 4>; L2 = Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >; Realm::DynamicTemplates::TagType = unsigned int]’:
/repo/runtime/realm/dynamic_templates.h:209:53:   required from ‘static void Realm::DynamicTemplates::IntDemuxHelper<TARGET, BASE, DELTA>::demux(int, T1, T2) [with T1 = unsigned int; T2 = int*; TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > > >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; int BASE = 1; int DELTA = 2]’
/repo/runtime/realm/dynamic_templates.h:211:71:   required from ‘static void Realm::DynamicTemplates::IntDemuxHelper<TARGET, BASE, DELTA>::demux(int, T1, T2) [with T1 = unsigned int; T2 = int*; TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > > >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; int BASE = 1; int DELTA = 3]’
/repo/runtime/realm/dynamic_templates.inl:154:64:   required from ‘static void Realm::DynamicTemplates::IntList<MIN, MAX>::demux(int, T1, T2) [with TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >  >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; T1 = unsigned int; T2 = int*; int MIN = 1; int MAX = 4]’
/repo/runtime/realm/dynamic_templates.inl:241:60:   required from ‘static void Realm::DynamicTemplates::ListProduct2<L1, L2>::demux(Realm::DynamicTemplates::TagType, A1) [with TARGET = Legion::Internal::NT_TemplateHelper::DimHelper; A1 = int*; L1 = Realm::DynamicTemplates::IntList<1, 4>; L2 = Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >; Realm::DynamicTemplates::TagType = unsigned int]’
/repo/runtime/legion/legion_types.h:1893:43:   required from here
```

(This repeats for quite a while.)

It doesn't look like an HTR error, but I'm not sure what else to say.

**Comment by lightsighter** [comment]
I'm confused. That is the same error from #1296. Are you tangling up your repositories? You shouldn't be getting that build error here.

**Comment by elliottslaughter** [comment]
With the fix from https://github.com/StanfordLegion/legion/issues/1296#issuecomment-1189676838, HTR now passes locally. (And yes, I need that patch in this branch or else my CI will not pass.)

I am now somewhat confused, because I would not have expected that to be the deciding factor in the tests passing. But I guess I'll see what CI says.

**Comment by elliottslaughter** [comment]
How strange. CI still fails: https://gitlab.com/StanfordLegion/legion/-/jobs/2744290515

I guess I'll try another replication attempt tomorrow, but I thought I already ran this configuration.

**Comment by elliottslaughter** [comment]
My ability to reproduce locally has been coming and going today, so it's been hard to work on this. But as best I can tell, this appears to be a case of LuaJIT's iteration order changing nondeterministically, creating downstream mismatches in the parsing of headers.

I have a tentative fix in HTR that I have sent to @mariodirenzo, that I'm waiting to test in CI to confirm it works.

**Comment by elliottslaughter** [comment]
I have confirmed my fix in CI, and am now waiting for @mariodirenzo for integration into the HTR CI branch.

**Comment by eddy16112** [comment]
I occasionally hit a realm crash in my developing branch, 
https://gitlab.com/StanfordLegion/legion/-/jobs/2754793811
Looks like it showed up before in the master branch
https://gitlab.com/StanfordLegion/legion/-/jobs/2746598895
It happens in shutdown.

**Comment by elliottslaughter** [comment]
@streichler will need to confirm, but I believe this is a preexisting issue.

**Comment by elliottslaughter** [comment]
This has been merged as of f5ca507dad83543eb1e62056a0ff26ea10b8c842.

@eddy16112, please be very careful when you update, because I aggressively rebased this branch. It would be best to completely remove any commits related to this branch if you can, or if absolutely necessary you can rebase. But if you rebase, please check that you get exactly the expected set of commits, because there were commits I deleted from this branch that I want to make sure do not reappear.
